fixes/c-oc-java (2018 edition)

aet opened the pull request:
New iteration of #6 with all changes re-applied manually against the latest master

Other than fixing random issues that I noticed against C/C++/Objective-C/Java sources, few other notable things:

- Tries to make language keyword maintenance a bit less error prone through inheritance
- C header files considered a new "language", .h usually can be a mixture of whatever C/C++/Objective-C
- Objective-C support modernization passes

Tested against 3 non-unity core source trees (Objective-C / Java) and fixed new regressions that have come since #6

No regressions against core unity source tree, fixes few formatting things

Few commits starting with (FIXED?) commit message seem to be fixed already in the upstream with some alternative way, but since the actual code changes still applied, I added them along the testcase.

The only code change entirely dropped from this PR is 01b65ac229d69f12a770dc5ba56214a0964b800b but the actual testcase is still included in 4c9ed33 to make sure it won't break later

Commit:
remove temp files (*.uncrustify) from tests/input/oc

Commit:
add a new base config for tests

with most whitespace related settings set to
something else than ignore, it's easier to
spot whitespace/type/cast issues

Commit:
improve CT_NEG handling with certain operators

Otherwise negative values could come up
with extra whitespace

Example failure from the diff file:

-    frames = ((frames + 15) & -16);
+    frames = ((frames + 15) & - 16);

Commit:
improve extern "C" / __attribute__ support

Otherwise types won't get properly detected and
with sp_arith=add, it leads to whitespace issues

Remove potential extra whitespace from extern "C"
or __attribute__ oneliners, there was no mathing
existing configuration option for such case and
unlikely need, too

This fixes the issue with C/CPP but not with OC/OC+,
which needs an extra commit for proper C keyword
support

Example failure from the diff file:

-extern "C" int *i;
+extern "C"    int *i;
 extern "C" {  int *i; }
 int *i;
-extern "C" FooString *i;
+extern "C"    FooString * i;
 extern "C" {  FooString *i; }
 FooString *i;

-__attribute__((visibility("default"))) int *i;
-__attribute__((visibility("default"))) FooString *i;
+__attribute__((visibility("default")))    int *i;
+__attribute__((visibility("default")))    FooString * i;

-#define DEFINE_NOTIFICATION(name) extern "C" __attribute__((visibility("default"))) FooString *const name = #name;
+#define DEFINE_NOTIFICATION(name) extern "C" __attribute__((visibility("default")))    FooString * const name = #name;

Commit:
fix sp_before_byref=remove on Java if statements with &&

Example failure from the diff file:

 public class JavaClass {
     public static void method() {
-        if (argA != null && argB != null) {
+        if (argA != null&& argB != null) {
         }
     }
 }

Commit:
improve nested generics with Java

">>>" symbol in punctuators.cpp is conflicting
with angle/template probe, causing incorrect
whitespace formatting

workaround by checking ">>>" explicitly

Example failure from the diff file:

 public class TestClass {
     private static void initMap(void) {
-        HashMap<String, HashMap<String, List<Track>>> resolutionTracks = new HashMap<String, HashMap<String, List<Track>>>();
+        HashMap < String, HashMap < String, List < Track >>> resolutionTracks = new HashMap < String, HashMap < String, List < Track >>> ();
     }

-    private static void addTrackToMap(String resolution, Track track, HashMap<String, HashMap<String, List<Track>>> resolutionTracks) {
+    private static void addTrackToMap(String resolution, Track track, HashMap < String, HashMap < String, List < Track >>> resolutionTracks) {
         HashMap<String, List<Track>> tracks = null;

Commit:
(FIXED?) fix an issue with Java generics and wildcards

a type like Map< ? , ? > can cause whitespace issues
outside the scope, causing sp_before_case_colon=remove
not to work correctly

Example failure from the diff file:

     private Map< ? , ? > map1 = null;
     private Map< ? , ? > map2 = null;
     private Map< ? , ? > map3 = null;

     public static HttpUriRequest getHttpUriRequest(TestClassAPIRequestMethod method, String apiPath) {
         switch (method) {
-        case BOTTOM_LEFT:
+        case BOTTOM_LEFT :
             break;
-        case GET:
+        case GET :
             req = new HttpGet(url);
             break;
-        case POST:
+        case POST :
             req = new HttpPost(url);
             break;
-        case PUT:
+        case PUT :
             req = new HttpPut(url);
             break;
-        case DELETE:
+        case DELETE :
             req = new HttpDelete(url);
             break;
         }

Commit:
fix whitespace issue with Java generics return type

Example failure from the diff file:

 public class TestClass {
-    public static <T> void executeRequest(final HttpUriRequest request, final ITestClassAPIResponseListener<T> responseListener) {
+    public static < T > void executeRequest(final HttpUriRequest request, final ITestClassAPIResponseListener<T> responseListener) {
     }

-    public <T extends YourType> T mymethod(T type) {
+    public < T extends YourType > T mymethod(T type) {
         return type;
     }
 }

Commit:
(FIXED?) fix sp_after_oc_msg_receiver=remove from breaking the source file

Example failure from the diff file:

 - (void)testMethod {
-    NSData *jsonData = [self dataUsingEncoding:NSUTF8StringEncoding];
-    id jsonParsedObj = [jsonSerializationClass JSONObjectWithData:jsonData options:0 error:&jsonError];
+    NSData *jsonData = [selfdataUsingEncoding:NSUTF8StringEncoding];
+    id jsonParsedObj = [jsonSerializationClassJSONObjectWithData:jsonData options:0 error:&jsonError];
     NSString *ret = [[TestClass sharedInstance]testString];
 }

Commit:
fix an indentation issue with CT_OC_BLOCK_EXPR

In certain cases, when mixing C/Obj-C calls and nesting
blocks internally, the indentation level fails

Nesting multiple blocks can also fail with the indentation level,
if the parent method calls arguments (CT_OC_MSG_NAME) are split
to multiple lines

Example failure from the diff file:

     [TestClassRequest performMethod:TestClassRequestMethodPOST
                     responseHandler:^(NSURLResponse *response) {
                         dispatch_after(retryTime, dispatch_get_main_queue(), ^(void) {
-                            [self postSelection:testName selection:selection];
-                        });
+            [self postSelection:testName selection:selection];
+        });
                     }];

     [TestClassRequest performMethod:TestClassRequestMethodPOST
                     responseHandler:^(NSURLResponse *response, NSData *responseData, NSError *error) {
                         [self checkPermissions:connection withCallback:^(NSError *error, id data) {
-                            [self bar];
-                        }];
+            [self bar];
+        }];
                     }];

Commit:
fix negative value support with OC @ literals

Example failure from the diff file:

-    NSNumber *foo1 = @-1;
+    NSNumber *foo1 = @ - 1;

Commit:
don't set CT_OC_MSG before type check

this fixes whitespace issues with certain [] cases
where we bail out after figuring out it's not an OC msg

Example failure from the diff file:

-#define NELEM(x) ((int) (sizeof(x) / sizeof((x)[0])))
+#define NELEM(x) ((int) (sizeof(x) / sizeof((x) [0])))

Commit:
fix OC bracket whitespace issue within for/in enumerator

Example failure from the diff file:

-for (UIWindow *testWindow in [[UIApplication sharedApplication] windows]) {
+for (UIWindow *testWindow in[[UIApplication sharedApplication] windows]) {

Commit:
fix chunk_ends_type to check virtual brace close

Example failure from the diff file:

         if (exportError != nil)
             return callback(exportError, nil);
-        NSDictionary *settings = [self getSettings];
+        NSDictionary * settings = [self getSettings];

Commit:
fix chunk_ends_type at the beginning of macros

Example failure from the diff file:

 #define TESTCLASS_SINGLETON_FOR_CLASS(classname, accessorname) \
-    static classname *shared##classname = nil;
+    static classname * shared##classname = nil;

 #define PUSH_CONTEXT(contextArg) \
-    EAGLContext *oldContext = [EAGLContext currentContext]; \
+    EAGLContext * oldContext = [EAGLContext currentContext]; \
     if (oldContext != contextArg) { \
         [EAGLContext setCurrentContext:contextArg]; \
     }

 #define CURRENT_CONTEXT \
-    NSOpenGLContext *currentContext = [NSOpenGLContext currentContext];
+    NSOpenGLContext * currentContext = [NSOpenGLContext currentContext];

Commit:
fix chunk_ends_type/fix_casts for special macros that use ##

Example failure from the diff file:

 #define MACRO(foo) \
     if (_##foo == NULL) { \
         Log("do %s", str(foo)); \
-        _##foo = (foo##_t *) process(PR_FLAG, str(foo)); \
+        _##foo = (foo##_t *)process(PR_FLAG, str(foo)); \
     }

 #define OVERLOAD(base, foo) \
-    foo##_override_t *foo##_bar = (foo##_override_t *) process(base##_bar, str(foo)); \
-    _##foo##_override = (foo##_override_t *) process(base##_cache, str(foo));
+    foo##_override_t * foo##_bar = (foo##_override_t *)process(base##_bar, str(foo)); \
+    _##foo##_override = (foo##_override_t *)process(base##_cache, str(foo));

Commit:
(FIXED?) fix OC protocol support within block literal

Example failure from the diff file:

 - (void)drawSomething:(id<MTLCommandBuffer>)commandBuffer {
-    [renderPass performDrawBlock:^(id<MTLRenderCommandEncoder> renderCommandEncoder) {
+    [renderPass performDrawBlock:^(id < MTLRenderCommandEncoder > renderCommandEncoder) {
         screenBlitObject.texture = src;

Commit:
fix OC msg send whitespace issue within preprocessor macros

Example failure from the diff file:

-#define UIColorFromRGB(rgbValue) [UIColor colorWithRed:((float) ((rgbValue & 0xFF0000) >> 16)) / 255.0 green:((float) ((rgbValue & 0xFF00) >> 8)) / 255.0 blue:((float) (rgbValue & 0xFF)) / 255.0 alpha:1.0]
+#define UIColorFromRGB(rgbValue) [UIColor colorWithRed: ((float) ((rgbValue & 0xFF0000) >> 16)) / 255.0 green: ((float) ((rgbValue & 0xFF00) >> 8)) / 255.0 blue: ((float) (rgbValue & 0xFF)) / 255.0 alpha: 1.0]

 #define NSLocalizedString(key, comment) \
-    [TestClassBundle localizedStringForKey:(key) value:@"" table:nil]
+    [TestClassBundle localizedStringForKey: (key) value: @"" table: nil]

Commit:
handle_oc_message_send: improve boxed expression support

Fixes whitespace issues when wrapping @()/@{}/@[] receivers
within Obj-C square brackets

Similar whitespace issue when wrapping the receiver within
parentheses, for example [(Foo *) (self.foo.bar) method]

Example failure from the diff file:

     headers = @{
-        @"Content-Length": [@(contentLength) stringValue],
+        @"Content-Length":[@(contentLength)stringValue],
         @"Content-Length2": [contentLength stringValue],
         @"Content-Disposition": [NSString stringWithFormat:@"name=\"%@\"; filename=\"%@\"", @"asset_data", identifier],
         @"Content-Type": @"application/octet-stream",

..

 - (BOOL)isStrimmed {
-    if ([(TestClassVideoEditorView *) (self.parentComponentsView.superview) isTrimmed]) {
+    if ([(TestClassVideoEditorView *) (self.parentComponentsView.superview)isTrimmed]) {
         return YES;
     }
     if ([((TestClassVideoEditorView *) self.parentComponentsView.superview) isTrimmed]) {

-    [menubutton.badge setBadgeText:[@(count + 1) stringValue]];
+    [menubutton.badge setBadgeText:[@(count + 1)stringValue]];

Commit:
improve type checking when overriding FUNC_DEF

This fixes whitespace issues with special macros that
could be used to construct entire functions or function
definitions

Example failure from the diff file:

-extern "C" void function1(void *self, uint32_t *arg2, uint32_t *arg3);
+extern "C" void function1(void *self, uint32_t * arg2, uint32_t * arg3);

-MACRO1(void, function2, type1 arg1, type2 arg2, const type1 *arg3);
+MACRO1(void, function2, type1 arg1, type2 arg2, const type1 * arg3);

 MACRO2(status_t, function3,
     void *arg1,
-    const sp<IFoo>& arg2) {
+    const sp<IFoo>&arg2) {
 }

-MACRO2(type4, function4, const void **arg1, type1 arg2, const type3 *arg3, type4 arg4, type4 arg5, bool arg6) {
+MACRO2(type4, function4, const void **arg1, type1 arg2, const type3 * arg3, type4 arg4, type4 arg5, bool arg6) {

Commit:
improve fix_casts for Java/Obj-C

Certain type casts (generics/protocols) cause whitespace
issues for both Java/Obj-C, with Java you can also refer to
a class with a member

Example failure from the diff file:

Java:

-            RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams) imageView.getLayoutParams();
+            RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams)imageView.getLayoutParams();

-        Map< ? , ? > map = (Map< ? , ? >) object;
+        Map< ? , ? > map = (Map< ? , ? >)object;

-        return (List<Object>) fromJSON((Object) obj);
+        return (List<Object>)fromJSON((Object) obj);

Obj-C:

-    previewViewController.previewControllerDelegate = (id<TestClassDelegate>) [TestClass sharedInstance];
+    previewViewController.previewControllerDelegate = (id<TestClassDelegate>)[TestClass sharedInstance];

Commit:
fix sp_after_angle issue on class definitions

When using sp_after_angle=remove and defining a class
with Obj-C protocols or Java generics that follow
BRACE_OPEN, it'll remove the whitespace between them

On normal class definitions without ANGLE_CLOSE, there
is no configuration available to choose between removing
or adding whitespace, basically just falling back to
default AV_ADD rule

This is potentially a case for a new configuration option,
but for now, let's make both cases identical with whitespace
formatting

Example failure from the diff file:

Java:

 public class TestClassPrefetchData implements ITestClassAPIInputStreamResponseListener {
-    private class TestClassPrefetchDataWrite extends AsyncTask<Void, Void, Void> {
+    private class TestClassPrefetchDataWrite extends AsyncTask<Void, Void, Void>{
     }
 }

Obj-C:

@interface TestClass2 : TestClassNavigationViewController {
}

-@interface TestClass : TestClassNavigationViewController<UIWebViewDelegate, UIActionSheetDelegate> {
+@interface TestClass : TestClassNavigationViewController<UIWebViewDelegate, UIActionSheetDelegate>{
    }

Commit:
Objective-C support modernization #1

.h header files are currently tokenized with both
C and C++ support, but when dealing with Objective-C
or Objective-C++ variants, you need to explicitly
set the language used for parsing

It's 2015 and Obj-C is pretty popular out there, it
shouldn't require a separate knowledge, especially
when dealing with larger cross-platform codebases,
to figure out which version of the C family languages
one header may be using

Without Obj-C being turned on, a typical Obj-C method
declaration in a header file gets considered as C cast
and then something. If the whitespace rules try to
match Apple style for OC method declarations, there
will be whitespace issues

A new flag FLAG_HDR is added, which could be potentially
used for something in the future. Right now it's only
being used to make "C-Header" unique from other potential
language flags that'd otherwise match the same mask, such
as "OC+" (Parsing: file.h as language C-Header)

Example failure from the diff file:

-- (void)print;
-- (void)setNumerator:(int)d;
-- (void)setDenominator:(int)d;
-- (int)numerator;
-- (int)denominator;
-- (void)setNumDen:(int)n:(int)d;
+- (void) print;
+- (void) setNumerator:(int) d;
+- (void) setDenominator:(int) d;
+- (int) numerator;
+- (int) denominator;
+- (void) setNumDen:(int) n:(int) d;

Commit:
Objective-C support modernization #2

When tokenizing source file, add a probe for checking
Obj-C support as a last resort before considering the
token as garbage

Add LANG_OC to all punctuators with LANG_C flag,
testcase would fail otherwise with CT_DC_MEMBER

Example failure from the diff file:

-#define TCH_ELOG TCH_LOGV(@"%s", __FUNCTION__)
+#define TCH_ELOG TCH_LOGV(@ "%s", __FUNCTION__)

-void TCH_initTestClass(UIViewController *viewController, id<TestClassDelegate> testclassDelegate) {
+void TCH_initTestClass(UIViewController *viewController, id < TestClassDelegate > testclassDelegate) {

-    id<TestClassDelegate> testclassDelegateRef = testclassDelegate;
+    id < TestClassDelegate > testclassDelegateRef = testclassDelegate;

-    [TCH_NS(TestClass) initWithDelegate:testclassDelegateRef andParentViewController:viewController];
+    [TCH_NS(TestClass) initWithDelegate: testclassDelegateRef andParentViewController: viewController];

-    ret = (bool) [TCH_NS(TestClass) isSupported];
+    ret = (bool)[TCH_NS(TestClass) isSupported];

Commit:
Objective-C support modernization #3

Assume that C++/Objective-C/Objective-C++ are always
a superset of C. While this may not always be true in real world,
for what uncrustify cares of, it does make maintaining generic
support for C family languages easier

This fixes C code support within Obj-C code, previously
keywords like 'extern', 'struct' or '__attribute__' were
not known for Obj-C, causing whitespace/indentation issues

- Add Obj-C variant of C testcase 09601

- Fix Obj-C testcase 50601 whitespace issue due to new keywords added

Other Obj-C related keyword changes:

- Improved C++/Objective-C++ interoperability, Obj-C 'self' is now CT_THIS

- '__block' is not restricted to Obj-C only where available, moved to LANG_C

- Added new Obj-C/ARC/C CoreFoundation bridging related qualifiers to fix
  potential whitespace issues:

  __autoreleasing, __bridge, __bridge_retained, __bridge_transfer,
  __strong, __typeof, __unsafe_unretained, __weak

- Add '@protected' and '@public' for Obj-C to match existing '@private'

C++:

C++ keyword 'export' / CT_EXPORT moved from LANG_C to LANG_CPP,
it's status changed to unused but reserved since C++11 spec

Example failure from the diff file:

-struct {
+struct
+{
     int foo;
 } bar;

-extern "C" int *i;
+extern "C"    int *i;
 extern "C" {  int *i; }
 int *i;
-extern "C" NSString *i;
+extern "C"    NSString * i;
 extern "C" {  NSString *i; }
 NSString *i;

-__attribute__((visibility("default"))) int *i;
-__attribute__((visibility("default"))) NSString *i;
+__attribute__((visibility("default")))    int *i;
+__attribute__((visibility("default")))    NSString * i;

-#define DEFINE_NOTIFICATION(name) extern "C" __attribute__((visibility("default"))) NSString *const name = @#name;
+#define DEFINE_NOTIFICATION(name) extern "C" __attribute__((visibility("default")))    NSString * const name = @#name;

Commit:
fix typeof whitespace issue

A bit tricky here, for unknown reason bunch of
keywords are handled as CT_SIZEOF, typeof variants
included.

Let's settle with adding CT_TYPE_CAST parent type
for CT_SIZEOF chunks and using it at sp_after_cast

Example failure from the diff file:

 - (void)method1 {
-    __weak __typeof(self) weakSelf1 = self;
-    __weak typeof(self) weakSelf2 = self;
+    __weak __typeof(self)weakSelf1 = self;
+    __weak typeof(self)weakSelf2 = self;
     __weak MyViewController *weakSelf3 = self;
 }

Commit:
add sp_inside_braces_oc_dict for @{} boxed expressions

this allows more fine-grained control of whitespace when
sp_inside_braces is still set to ignore

Example failure from the diff file:

-    NSDictionary *productParameters = @{ @"id": appStoreID };
-    NSDictionary *options = @{ AVURLAssetPreferPreciseDurationAndTimingKey: @YES };
-    NSDictionary *baz = @{ foo: bar };
+    NSDictionary *productParameters = @{   @"id": appStoreID   };
+    NSDictionary *options = @{   AVURLAssetPreferPreciseDurationAndTimingKey: @YES   };
+    NSDictionary *baz = @{foo: bar};

Commit:
add sp_func_call_user_inside_fparen/sp_func_call_user_paren_paren

by extending existing set FUNC_CALL_USER behaviour, you
can now use it with special macros that take a code
expression as an argument and maintain whitespace
padding for extra readability

Example failure from the diff file:

-    CMD_CHECK( fflush(stdout); );
-    CMD_CHECK( fflush(stdout) );
+    CMD_CHECK(fflush(stdout););
+    CMD_CHECK(fflush(stdout));

Commit:
fix chunk_ends_type near CT_PP_ types

Example failure from the diff file:

 #else
-        NSString *extension = [thumbnailUrl pathExtension];
+        NSString * extension = [thumbnailUrl pathExtension];

..

 #if DEBUG
-                            NSString *rsp = [[NSString alloc] initWithData:responseData encoding:NSUTF8StringEncoding];
+                            NSString * rsp = [[NSString alloc] initWithData:responseData encoding:NSUTF8StringEncoding];
                             TestClassLog(@"TOGGLE CONNECTION ADDED response:%li responseData:%@ error:%@", (long) [((NSHTTPURLResponse *) response) statusCode], rsp, [error localizedDescription]);
 #endif
-                            NSJSONSerialization *jsonConnection = [responseData TestClassJSONObject];
+                            NSJSONSerialization * jsonConnection = [responseData TestClassJSONObject];

Commit:
add OC test method_ends_semicolon.m

(was: improve CT_OC_MSG_DECL type check)

Occasionally there will be user errors where someone will
copy the interface method declaration to implementation
and leaves the semicolon

Depending on brace settings, this leads to incorrect braces handling

Example failure from the diff file:

-+ (void)cancelRequest:(id)request; {
++ (void)cancelRequest:(id)request;
+{
 }

Commit:
fix a typo

Commit:
allow macros to have unmatched BRACE_CLOSE

Fixes testsuite hard error

812:   Uncrustify error: Parsing: input/oc/macro-close-brace.m as language OC
812:
812:   Unmatched BRACE_CLOSE
812:
812:   at orig_line=18, orig_col=29
812:
812:   Try the option 'tok_split_gte = true'

Commit:
fix CT_BOOL regression from 59fd03e966e23b6a2bbe1b599a9e1bc1484ad485

in the current master, it seems this change is no longer required
to fix issue #548 along the testcase for it (tests/input/cpp/bug_33056.cpp)

fixes regression with:

 #import <Foundation/Foundation.h>

-#if TARGET_OS_MAC && !TARGET_OS_IPHONE
+#if TARGET_OS_MAC&& !TARGET_OS_IPHONE
 #import <AppKit/AppKit.h>
 #endif

Commit:
downgrade objective-c CT_THIS into CT_WORD in some cross-platform cases

follow-up to fdbcd261ca5875e837bf659663bedb35cef05df9

Commit:
fix formatting with obj-c and react native RCT_EXPORT_METHOD macro

Commit:
improve 24e0d6dd3f3d8aeb486e3ec1c3cff8b80938f60a / issue# 1276

Example failure from the diff file:

-    [TestCase("", new string[0])]
-    [TestCase(";", new string[0])]
+    [TestCase("", new string [0])]
+    [TestCase(";", new string [0])]
     public void SetFlags_iOS(string flags,  string[] expected)

Commit:
fix java annotation issue with extra newline

Example failure from the diff file:

-    @Test(groups = {"testdroid", "local"})
+    @Test(groups = {"testdroid", "local"}
+          )

Commit:
improve 24e0d6dd3f3d8aeb486e3ec1c3cff8b80938f60a / issue# 1266

Example failure from the diff file:

-#define FOO(x) (((uintptr_t) (x) & 0x1) == 0)
+#define FOO(x) (((uintptr_t) (x)& 0x1) == 0)

 static uint8_t bar(void **ptr) {
 }

Commit:
update cli tests output results

mihaipopescu commented on src/keywords.cpp:
```diff
@@ -47,287 +47,342 @@ static int kw_compare(const void *p1, const void *p2);
+ * of C, but since they have been previously defined for LANG_C,
+ * let's not break existing behaviour
+ *
+ * Type inheritance is implemented at init_keywords()
```
💯

mihaipopescu commented on src/punctuators.cpp:
```diff
@@ -26,97 +26,97 @@ using namespace std;
+   { R"_(??<)_", CT_BRACE_OPEN,   LANG_C | LANG_CPP | LANG_OC | FLAG_DIG                                 }, // trigraph {
+   { R"_(??>)_", CT_BRACE_CLOSE,  LANG_C | LANG_CPP | LANG_OC | FLAG_DIG                                 }, // trigraph }
+   { R"_(??-)_", CT_INV,          LANG_C | LANG_CPP | LANG_OC | FLAG_DIG                                 }, // trigraph ~
+   { R"_(??!)_", CT_ARITH,        LANG_C | LANG_CPP | LANG_OC | FLAG_DIG                                 }, // trigraph |
```
could've used the same language keywords inheritance concept here as well but these are fewer.

aet replied:
True, I also noticed that I had added a duplicate entry for LANG_OC into CT_ELLIPSIS, fixed in a95b4e8b

mihaipopescu commented on src/uncrustify.cpp:
```diff
@@ -2099,17 +2101,18 @@ struct lang_name_t
+   { "OC+",      LANG_OC | LANG_CPP            },
+   { "CS+",      LANG_CS | LANG_CPP            },
+   { "ECMA",     LANG_ECMA                     },
+   { "C-Header", LANG_OC | LANG_CPP | FLAG_HDR },
```
Awesome 💯. We actually had a hack in our code to treat all header files as `OC+` as there was no way to tell them apart. I think this change needs to be documented in the PR description. cc @Rassi

mihaipopescu reviewed (approved):
I really welcome this long waited PR.

Along with all Objective-C fixes, it introduces two novelties that I like:
1. language keyword inheritance
2. new "language" for c header files

Commit:
remove duplicate LANG_OC from CT_ELLIPSIS

CDanU commented on src/space.cpp:
```diff
@@ -843,8 +843,11 @@ static argval_t do_space(chunk_t *first, chunk_t *second, int &min_sp, bool comp
+      if (first->type != CT_QUALIFIER)
+      {
+         log_rule("sp_before_angle");
+         return(cpd.settings[UO_sp_before_angle].a);
```
Why is this if check necessary?

mihaipopescu replied:
Is the change accompanied by a test?

CDanU replied:
Yes, 018e14b9e24e9f5d8bb88b0d17ec83fa13db299d .. well kind of.

I am not sure why the option shouldn't also be applied if there is a `CT_QUALIFIER` before the generic `<`. Java should still compile if `remove` is used to generate something like `public static<T> void` ...

aet reviewed (commented):
commit  9cc01c8 has a wrong commit id in the message, it should be e1e449f34

the issue tracking number is correct though

Commit:
'__block' is not restricted to Obj-C only where available, moved to LANG_C

this was originally supposed to be already in fdbcd261ca5875e837bf659663bedb35cef05df9,
but slipped while re-applying changes against the latest codebase

mihaipopescu commented:
jenkins builder: test this please
